### PR TITLE
TeraEd: Run Terasology in Swing/AWT

### PR DESCRIPTION
**Pull request by hajdam**
### Contains

Run Terasology in TeraEd

terasology-teraed

### How to test

Run Terasology and check if it works unchanged (I didn't run much tests)
Run TeraEd and check if Terasology starts in it (currently only game menu works)

### Notes

This is first small step to make TeraEd do something, but I actually doubt this will be merged as it makes a bit of a (additional) mess in the code.

Changes modify rendering methods, because for the openGL to run in canvas, you need to run all opengl operations during canvas render instead of typical at any time assuming GL window exist. Also swing rendering thread needs to be used instead of main loop.

If approved, more functionality could be added to TeraEd. Feel free to recommend any improvements or if anyone know how to make it load a save/new game, let me know...

### CODE REVIEW

File: facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortlet.java
Comment (DarkWeird) [inline facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortlet.java:138]: Hmm.
there should runs all render code, isn't it?
Seems, it is best place for rendering ticks

File: facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortlet.java
Comment (DarkWeird) [inline facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortlet.java:203]: Swing's key codes equal glfw's key codes?

File: facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortletDisplayDevice.java
Comment (DarkWeird) [inline facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortletDisplayDevice.java:41]: This data can be changed by resize?

File: facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortletDisplayDevice.java
Comment (DarkWeird) [inline facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortletDisplayDevice.java:105]: Commented code should be eliminated

File: facades/TeraEd/src/main/java/org/terasology/editor/ui/Viewport.java
Comment (DarkWeird) [inline facades/TeraEd/src/main/java/org/terasology/editor/ui/Viewport.java:40]: Double line

File: facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortlet.java
Comment (hajdam) [reply facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortlet.java:203]: Partially. Characters equals directly, special keys will have to be converted. Didn't do this yet. Same applies to mouse events - idea here is, that there will be for example mode in editor, where you click on it and mouse will go to exclusive mode to test gameplay until ESC for example... Or something like that.

File: facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortletDisplayDevice.java
Comment (hajdam) [reply facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortletDisplayDevice.java:41]: The reason for this is, that it calls GL11 functions which requires active OpenGL context to process, so have to be done in display thread. There are currently vendor strings, so it should be ok to call once (unless someone can switch graphic cards in the middle of playing?)

File: engine/src/main/java/org/terasology/engine/subsystem/headless/HeadlessGraphics.java
Comment (hajdam) [reply engine/src/main/java/org/terasology/engine/subsystem/headless/HeadlessGraphics.java:78]: Yes, but I had to extract openGL initialization as it has to be done with active OpenGL, which is not the case for TeraEd (subsystem constructors are run before TeraEd have got OpenGL canvas available - glInitialize is postponed for later call in render method there)

File: engine/src/main/java/org/terasology/input/lwjgl/LwjglKeyboardDevice.java
Comment (hajdam) [reply engine/src/main/java/org/terasology/input/lwjgl/LwjglKeyboardDevice.java:176]: I though I needed it there, but I can move it out. Well, it's not nice that keyboard registers itself to GL window as side effect in constructor anyway...

File: engine/src/main/java/org/terasology/engine/subsystem/DisplayDeviceInfo.java
Comment (hajdam) [reply engine/src/main/java/org/terasology/engine/subsystem/DisplayDeviceInfo.java:6]: Well, it is actually already the case as there is HeadlessDisplayDevice. Didn't want to tackle that as it is currently used for metrics only. But I'm ok with any alternative you wish to suggest.

File: facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortletDisplayDevice.java
Comment (hajdam) [reply facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortletDisplayDevice.java:105]: Forgot to remove that :-(

File: engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphicsManager.java
Comment (hajdam) [reply engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphicsManager.java:74]: Sorry, I don't understand what you mean here... This variable here is to indicate if graphics operations can be performed in game thread (PC) or must be executed in display thread later (TeraEd).

File: facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortletDisplayDevice.java
Comment (DarkWeird) [reply facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortletDisplayDevice.java:41]: Hmm it maybe interesting- change graphics during teraed session. But not in resize event.

File: engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphicsManager.java
Comment (DarkWeird) [reply engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphicsManager.java:74]: I thought that is rendering separation in engine on rendering and logic thread.

Comment (hajdam) [reply engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphicsManager.java:74]: I did think so too, but I hadn't find it (yet).

File: facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortletDisplayDevice.java
Comment (hajdam) [reply facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortletDisplayDevice.java:41]: Well, it could be red every render tick, so that it will be available for metrics 1-tick delayed in the worst case.

Comment (DarkWeird) [reply facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortletDisplayDevice.java]: Ideally render ticks should works every 1/60 seconds.
then 1-tick delay is not afraid :)

File: facades/TeraEd/src/main/java/org/terasology/editor/TeraEd.java
Comment (DarkWeird) [inline facades/TeraEd/src/main/java/org/terasology/editor/TeraEd.java:69]: Your game not runs in teraEd because LwjglTime returns 0.0 every time (see StageLoading)
You should initialize GLFW for timer starts counts. (or use different timer)

```suggestion

            if (!GLFW.glfwInit()) {
                throw new RuntimeException("Failed to initialize GLFW");
            }
            sceneProperties = new SceneProperties(engine);
```

**Review by DarkWeird — COMMENTED** [review]
AWT-Event-Queue as main thread feels wrong T_T. (but I suspect it is single possible way, isn't it?)

(after GLFW.init suggestion applying)
Also something corrupted with tick lifecycle(recursion via postUpdate-render-prepareToRender-paintGL-tick-postUpdate:
```
at com.badlogic.gdx.physics.bullet.BulletBase.<init>(BulletBase.java:34)
	at com.badlogic.gdx.physics.bullet.collision.btCollisionWorld.<init>(btCollisionWorld.java:23)
	at com.badlogic.gdx.physics.bullet.collision.btCollisionWorld.<init>(btCollisionWorld.java:29)
	at com.badlogic.gdx.physics.bullet.dynamics.btDiscreteDynamicsWorld.getCollisionWorld(btDiscreteDynamicsWorld.java:138)
	at org.terasology.physics.bullet.BulletPhysics.getNewCollisionPairs(BulletPhysics.java:751)
	at org.terasology.physics.bullet.BulletPhysics.update(BulletPhysics.java:294)
	at org.terasology.physics.engine.PhysicsSystem.update(PhysicsSystem.java:185)
	at org.terasology.engine.modes.StateIngame.update(StateIngame.java:171)
	at org.terasology.engine.TerasologyEngine.tick(TerasologyEngine.java:496)
	at org.terasology.editor.subsystem.LwjglPortlet$1.paintGL(LwjglPortlet.java:131)
	at org.terasology.editor.subsystem.LwjglPortletDisplayDevice.prepareToRender(LwjglPortletDisplayDevice.java:110)
	at org.terasology.engine.modes.StateIngame.render(StateIngame.java:202)
	at org.terasology.editor.subsystem.LwjglPortlet.postUpdate(LwjglPortlet.java:95)
	at org.terasology.engine.TerasologyEngine.tick(TerasologyEngine.java:505)
	at org.terasology.editor.subsystem.LwjglPortlet$1.paintGL(LwjglPortlet.java:131)
	at org.terasology.editor.subsystem.LwjglPortletDisplayDevice.prepareToRender(LwjglPortletDisplayDevice.java:110)
	at org.terasology.engine.modes.StateIngame.render(StateIngame.java:202)
	at org.terasology.editor.subsystem.LwjglPortlet.postUpdate(LwjglPortlet.java:95)
	at org.terasology.engine.TerasologyEngine.tick(TerasologyEngine.java:505)
	at org.terasology.editor.subsystem.LwjglPortlet$1.paintGL(LwjglPortlet.java:131)
	at org.terasology.editor.subsystem.LwjglPortletDisplayDevice.prepareToRender(LwjglPortletDisplayDevice.java:110)
	at org.terasology.engine.modes.StateIngame.render(StateIngame.java:202)
	at org.terasology.editor.subsystem.LwjglPortlet.postUpdate(LwjglPortlet.java:95)
	```

File: facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortletDisplayDevice.java
Comment (DarkWeird) [inline facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortletDisplayDevice.java:99]: You don't need this.
You don't use GLFW's window

File: facades/TeraEd/src/main/java/org/terasology/editor/input/AwtMouseDevice.java
Comment (DarkWeird) [inline facades/TeraEd/src/main/java/org/terasology/editor/input/AwtMouseDevice.java:138]: see:
https://stackoverflow.com/questions/32159065/how-to-grab-a-mouse-in-java-swing

Comment (hajdam) [reply facades/TeraEd/src/main/java/org/terasology/editor/input/AwtMouseDevice.java:138]: This will require additional logic for shifting between regular and grab mode. There is also issue with rotating already.

**Comment by hajdam** [comment]
I failed at all attempts to make this work beyond rendering of the game menu as there seems to be some issue with order of initialization. I think it would require deeper separation of rendering and main thread or different approach. Blocks are visible in game after graphical settings is changed, but there are still other issues...

But if interested, I think it might be still ok to include this if it won't break/slows main game and make it work later.

**Comment by pollend** [comment]
@hajdam that's more function then what is currently in the eninge. not too bad to have this working again if only in some basic capacity. I do remember this coming up that might be relevant to this effort.

https://github.com/JetBrains/compose-jb

**Comment by DarkWeird** [comment]
Calling refreshing DAG helps(i did it in debug). World rendering began renders right.
 Idk why...
Dag refreshing at start and at change settings.
It is happens in sane thread. (Via WorldRenderer#render)

@dave2s can you help? Any ideas?

@hajdam TeraPC works. I think we can merge this PR when black screen in game will resolved. Thx for TeraED restoring!

**Comment by hajdam** [comment]
Still no luck
terasology-teraed-broken

**Comment by hajdam** [comment]
After few painful hours, hopefully it will work at least a bit. Please test this on Mac OSX.

File: engine/src/main/java/org/terasology/rendering/opengl/GLSLMaterial.java
Comment (DarkWeird) [inline engine/src/main/java/org/terasology/rendering/opengl/GLSLMaterial.java]: commented code

**Review by DarkWeird — DISMISSED** [review]
Great work!

(linux - works)

outstanding - check running TeraPC and TeraEd on macos.
@Cervator 
@skaldarnar 

I don't think that is blocking for merging.

File: engine/src/main/java/org/terasology/rendering/opengl/GLSLMaterial.java
Comment (hajdam) [reply engine/src/main/java/org/terasology/rendering/opengl/GLSLMaterial.java]: Fixed in following commit.

Comment (DarkWeird) [reply engine/src/main/java/org/terasology/rendering/opengl/GLSLMaterial.java]: oh, ok. 
I reviewed commit 5 only :D

**Comment by Cervator** [comment]
Nice work @hajdam ! 

I tested on Mac but did not get very far. I adjusted the TeraEd run config to include the unique parameters we need for Macs, but still get a very Mac-sounding error on trying to start. I imagine @DarkWeird did something to the PC facade as part of the LWJGL3 upgrade that we might need to replicate for TeraEd

```
/Users/cervator/dev/zulu11.31.11-ca-fx-jdk11.0.3-macosx_x64/bin/java -Xms256m -Xmx1536m -XstartOnFirstThread -Djava.awt.headless=true -javaagent:/Applications/IntelliJ IDEA CE.app/Contents/lib/idea_rt.jar=64504:/Applications/IntelliJ IDEA CE.app/Contents/bin -Dfile.encoding=UTF-8 -classpath /private/var/folders/j9/9d141hz563x0qdq1098_92100000gp/T/classpath157599996.jar org.terasology.editor.TeraEd -homedir -noCrashReport -noSplash
Exception in thread "main" java.awt.HeadlessException
	at java.desktop/java.awt.GraphicsEnvironment.checkHeadless(GraphicsEnvironment.java:208)
	at java.desktop/java.awt.Window.<init>(Window.java:548)
	at java.desktop/java.awt.Frame.<init>(Frame.java:423)
	at java.desktop/java.awt.Frame.<init>(Frame.java:388)
	at java.desktop/javax.swing.SwingUtilities$SharedOwnerFrame.<init>(SwingUtilities.java:1919)
	at java.desktop/javax.swing.SwingUtilities.getSharedOwnerFrame(SwingUtilities.java:1995)
	at java.desktop/javax.swing.JWindow.<init>(JWindow.java:183)
	at java.desktop/javax.swing.JWindow.<init>(JWindow.java:135)
	at org.terasology.editor.TeraEd.<init>(TeraEd.java:33)
	at org.terasology.editor.TeraEd.main(TeraEd.java:42)
```

Specifically the Mac-specific params are `-XstartOnFirstThread -Djava.awt.headless=true` and it is probably that headless AWT thing we're running into. We needed that for the main PC facade for whatever reason, maybe @DarkWeird can expand on that, maybe on Discord sometime?

It might be something easy - and I hope so, since it can be tricky to deal with Mac specific issues, especially when it about takes a rendering wizard to experiment :-)

On a different note are all the changes in this PR legitimately related to just TeraEd? It looks to me like there's a _bunch_ more stuff including auto-formatting, type fixes, and a ton of LWJGL stuff that makes me wonder if the LWJGL3 upgrade commit somehow got included in the diff over some quirk.

**Comment by DarkWeird** [comment]
Yeah.
 '-XstartOnFirstThread' required.
Headless not needs( i hope at least, because we use awt on teraED in headed mode)
Just another run config fir TeraED.

Lwjgl3 code from engine was refactored legally :D
Extracted some common parts. DRY. (And also formating)

**Comment by Cervator** [comment]
Oho! Thanks for that, both the confirmation of changes and the headless note. Taking off that flag did get me further, I guess we needed it specifically for the regular game, but now specifically not for TeraEd.

Now it just arbitrarily stops after this point:


```
/Users/cervator/dev/zulu11.31.11-ca-fx-jdk11.0.3-macosx_x64/bin/java -Xms256m -Xmx1536m -XstartOnFirstThread -javaagent:/Applications/IntelliJ IDEA CE.app/Contents/lib/idea_rt.jar=65016:/Applications/IntelliJ IDEA CE.app/Contents/bin -Dfile.encoding=UTF-8 -classpath /private/var/folders/j9/9d141hz563x0qdq1098_92100000gp/T/classpath381416710.jar org.terasology.editor.TeraEd -homedir -noCrashReport -noSplash
PathManager being initialized. Initial code location is /Users/cervator/dev/terasology/git_ws/mini/Terasology/engine/build/classes
Switched it to expected working dir: /Users/cervator/dev/terasology/git_ws/mini/Terasology
installPath was null, running from IDE or headless server? Setting to: /Users/cervator/dev/terasology/git_ws/mini/Terasology
Found the natives dir: /Users/cervator/dev/terasology/git_ws/mini/Terasology/natives
15:09:28.214 [main] INFO  c.b.gdx.physics.bullet.NativeSupport - Loading JNBullet Library:bullet
15:09:28.249 [main] INFO  c.b.gdx.physics.bullet.NativeSupport - 	Loaded from java.library.path: /Users/cervator/dev/terasology/git_ws/mini/Terasology/natives/macosx/libbullet-darwin-amd64.dylib
```

At this point a little exec window named `java` has shown up, but there's seemingly no way to make it visible.

Still, considering that TeraEd doesn't work anyway, it seems like we should probably just merge this as-is? Even if TeraEd remains broken on Mac that doesn't seem like a huge loss.

**Comment by hajdam** [comment]
Hopefully properly merged. Am I suppose to perform merges to keep the PR up to date? Or was it already denied / not accepted, or is it just missing something or people are just busy...?

**Comment by Cervator** [comment]
> Hopefully properly merged. Am I suppose to perform merges to keep the PR up to date? Or was it already denied / not accepted, or is it just missing something or people are just busy...?

Just busy! Plus we were trying to stabilize a few things over the holidays to see if we could get a good build, before merging in tons more stuff. That's past now and I've got this sitting in a browser tab myself for merging soon ™️ if nobody else beats me to it :-)

**Review by DarkWeird — APPROVED** [review]
Imho, macos is nit critical for TeraED.
I don't want that pr will stale again and again.

**Pull request merged**
